skyportal/tests/frontend/test_sources.py

```diff
@@ -84,3 +84,16 @@ def test_download_comment_attachment(driver, user, public_source):
         assert l.split('\n')[0] == 'wavelength,flux,instrument_id'
     finally:
         os.remove(fpath)
+
+
+
+def test_view_only_user_cannot_comment(driver, view_only_user, public_source):
+    driver.get(f"/become_user/{view_only_user.id}")
+    driver.get(f"/source/{public_source.id}")
+    driver.wait_for_xpath(f'//div[text()="{public_source.id}"]')
+    try:
+        driver.wait_for_xpath('//input[@name="comment"]', timeout=0)
+    except:
+        pass
+    else:
+        raise Exception('Comment form is visible to view-only user, but should be hidden.')
```

Comment:
Nicely done.

This seems to become a pattern, so we can also add a little shim: `driver.assert_missing_xpath(XPATH, error message)`?  Will leave that for another PR.

Reply:
Good call - after this one gets merged, I'll add that to baselayer and update all the skyportal tests (including this one) accordingly.